R/md_document.R

```diff
@@ -88,3 +92,41 @@ md_document <- function(variant = "markdown_strict",
     post_processor = post_processor
   )
 }
+
+adapt_md_variant <- function(variant, preserve_yaml) {
+  variant_base <- gsub("^([^+-]*).*", "\\1", variant)
+  variant_extensions <- gsub(sprintf("^%s", variant_base), "", variant)
+
+  set_extension <- function(format, ext, add = TRUE) {
+    ext <- paste0(ifelse(add, "+", "-"), ext)
+    if (grepl(ext, format, fixed = TRUE)) return(format)
+    paste0(format, ext, collapse = "")
+  }
+
+  add_yaml_block_ext <- function(extensions, preserve_yaml) {
+    set_extension(variant_extensions, "yaml_metadata_block", preserve_yaml)
+  }
+
+  # yaml_metadata_block extension
+  variant_extensions <- switch(
+    variant_base,
+    gfm =,
+    commonmark =,
+    commonmark_x = {
+      if (pandoc_available(2.13)) {
+        add_yaml_block_ext(variant_extensions, preserve_yaml)
+      } else {
+        variant_extensions
+      }
+    },
+    markdown =,
+    markdown_phpextra =,
+    markdown_github =,
+    markdown_mmd =,
+    markdown_strict = add_yaml_block_ext(variant_extensions, preserve_yaml),
+    # do not modified for unknown (yet) md variant
+    variant_extensions
+  )
```

Comment:
Below is how I would write the code (replacing `switch` with `if` to avoid the repetition of `add_yaml_block_ext(variant_extensions, preserve_yaml)`, and listing two conditions instead of enumerating all possible formats that could be left to be considered in the future):

```suggestion
  set_extension <- function(format, add = TRUE, ext = "yaml_metadata_block") {
    ext <- paste0(ifelse(add, "+", "-"), ext)
    if (grepl(ext, format, fixed = TRUE)) return(format)
    paste0(format, ext, collapse = "")
  }

  if ((variant_base == 'markdown_strict') ||
      (variant_base == 'commonmark_x' && pandoc_available(2.13))) {
    variant_extensions = set_extension(variant_extensions, preserve_yaml)
  }
```

It's just a suggestion. If you prefer `switch`, that's okay to me.

Reply:
Let me clarify something. We need 2 things: 

1. `add_yaml_block_ext(variant_extensions, preserve_yaml)` should be apply for each of those formats `markdown`,`markdown_phpextra`,`markdown_github`,`markdown_mmd`,`markdown_strict`. These are all the formats that supports the extension since a long time.
1. `if (pandoc_available(2.13)) add_yaml_block_ext(variant_extensions, preserve_yaml) else  variant_extensions}` should be apply for each of those formats:  `gfm`,`commonmark`,`commonmark_x`. That are the new format that supports the extensions since recently. 

This is why I used `switch()` to avoid a `if()` with long condition, as it offers a special syntax to list several cases and apply the same value

``` r
switch ('a',
    a = ,
    b = "I am a and b"
)
#> [1] "I am a and b"
```

I find it clearer to read to list all the formats also. 

Otherwise it would be 

```suggestion
  set_extension <- function(format, ext, add = TRUE) {
    ext <- paste0(ifelse(add, "+", "-"), ext)
    if (grepl(ext, format, fixed = TRUE)) return(format)
    paste0(format, ext, collapse = "")
  }

  add_yaml_block_ext <- function(extensions, preserve_yaml) {
    set_extension(variant_extensions, "yaml_metadata_block", preserve_yaml)
  }

  # yaml_metadata_block extension
  if ((variant_base %in% c('markdown', 'markdown_phpextra', 'markdown_github', 'markdown_mmd', 'markdown_strict')) || 
      (variant_base %in% c('gfm', 'commonmark', 'commonmark_x') && pandoc_available('2.13'))) {
    variant_extensions = add_yaml_block_ext(variant_extensions, preserve_yaml)
  }
```

This would be longer lines but less lines, and less easier to read than the switch that allows nice reading of supported format with the indentation. However, it repeat the function `add_yaml_block_ext()` indeed. 

In your suggestion, if `variant_base == "gfm"` then, `variant_extensions = set_extension(variant_extensions, preserve_yaml)` would not be apply but we want it to be applied. am I missing something ? 

Also, I would like to keep `set_extension()` with no default, as I think this could be useful in other places.

Reply:
I see. I've never used `switch(x, a =, b = value)` before, so I misunderstood it. Sorry about that.

Reply:
No worries. I am glad you can also learn from me 😉 😄